components/renku_data_services/migrations/versions/584598f3b769_expand_and_separate_environments_from_.py

```diff
@@ -26,22 +26,48 @@
 
 def upgrade() -> None:
     # ### commands auto generated by Alembic - please adjust! ###
-    op.execute("DELETE FROM sessions.launchers")
-    op.drop_column("launchers", "default_url", schema="sessions")
-    op.drop_column("launchers", "environment_kind", schema="sessions")
-    op.drop_column("launchers", "container_image", schema="sessions")
+    # Migrate session launchers
     op.execute("DROP TYPE environmentkind CASCADE")
     op.execute("CREATE TYPE environmentkind AS ENUM ('GLOBAL', 'CUSTOM')")
-    op.add_column("environments", sa.Column("port", sa.Integer(), nullable=True), schema="sessions")
-    op.add_column("environments", sa.Column("working_directory", sa.String(), nullable=True), schema="sessions")
-    op.add_column("environments", sa.Column("mount_directory", sa.String(), nullable=True), schema="sessions")
-    op.add_column("environments", sa.Column("uid", sa.Integer(), nullable=True), schema="sessions")
-    op.add_column("environments", sa.Column("gid", sa.Integer(), nullable=True), schema="sessions")
     op.add_column(
         "environments",
         sa.Column("environment_kind", sa.Enum("GLOBAL", "CUSTOM", name="environmentkind"), nullable=True),
         schema="sessions",
     )
+    op.execute("UPDATE sessions.environments SET environment_kind = 'GLOBAL' WHERE environment_kind is NULL")
+    # NOTE: When the session launcher has environment_id set to null then it is a custom environment
+    # NOTE: We populate the name column in the new environment with the session launcher id
+    # This way we can join the newly add environments with their launchers and update the foreign key in the launchers
+    # with the newly created environments.
+    # NOTE: Since postgres cannot autogenerate ulids we use the session launcher ulid when creating the environment
+    op.execute(
+        "INSERT INTO sessions.environments(id, name, created_by_id, creation_date, container_image, default_url, environment_kind) "
+        "SELECT generate_ulid(), id, created_by_id, creation_date, container_image, default_url, 'CUSTOM' "
```

Comment:
issue: actually this would fail in prod, no? because the generate_ulid function is only available in the devcontainer, so locally and in CI?

Reply:
I added it to the db init in the renku repo PR: https://github.com/SwissDataScienceCenter/renku/pull/3712